ee/allowedcmd/run_disclaimed_darwin.go

```diff
@@ -0,0 +1,174 @@
+        return err;
+    }
+
+    pid_t pid;
+    // passing NULL for file_actions, we capture stderr/out from running this as subcommand
+    err = posix_spawn(&pid, path, NULL, &attrs, argv, envp);
+    posix_spawnattr_destroy(&attrs);
+    if (err != 0) {
+        return err;
+    }
+
+    int status;
+    if (waitpid(pid, &status, 0) == -1) {
+        return -1;
+    }
+
+    // WIFEXITED will return non-zero if the child process terminated normally
+    return WIFEXITED(status) ? WEXITSTATUS(status) : -1;
+}
+*/
+import "C"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os/user"
+	"unsafe"
+
+	"github.com/kolide/launcher/pkg/log/multislogger"
+)
+
+type allowedCmdGenerator struct {
+	allowedOpts map[string]struct{}
+	generate    func(ctx context.Context, args []string) (*TracedCmd, error)
+}
+
+func RunDisclaimed(_ *multislogger.MultiSlogger, args []string) error {
+	ctx := context.Background()
+	cmd, err := commandToDisclaim(ctx, args)
+	if err != nil {
+		return fmt.Errorf("gathering subcommand: %w", err)
+	}
+
+	cmdPath := C.CString(cmd.Path)
+	defer C.free(unsafe.Pointer(cmdPath))
+
+	// we're building up the C arguments here like []*C.char{cmdPath, arg1, arg2..., nil}
+	// note that cmdPath will already be present as the first argument
+	cargs := make([]*C.char, len(cmd.Args)+1) // 1 extra for null terminator
+	for i, arg := range cmd.Args {
+		cargs[i] = C.CString(arg)
+	}
+	cargs[len(cargs)-1] = nil // should be terminated by null entry
+
+	defer func() {
+		for _, arg := range cargs {
+			if arg != nil {
+				C.free(unsafe.Pointer(arg))
+			}
+		}
+	}()
+
+	cenvs := make([]*C.char, len(cmd.Environ())+1) // 1 extra for null terminator
+	for i, arg := range cmd.Environ() {
+		cenvs[i] = C.CString(arg)
+	}
+	cenvs[len(cenvs)-1] = nil // should be terminated by null entry
+
+	defer func() {
+		for _, arg := range cenvs {
+			if arg != nil {
+				C.free(unsafe.Pointer(arg))
+			}
+		}
+	}()
+
+	ret := C.spawn_disclaimed(cmdPath, &cargs[0], &cenvs[0])
+	if int(ret) != 0 {
+		return fmt.Errorf("spawn_disclaimed returned error code %d", int(ret))
+	}
+
+	return nil
+}
+
+func commandToDisclaim(ctx context.Context, args []string) (*TracedCmd, error) {
+	if len(args) < 1 {
+		return nil, errors.New("rundisclaimed expects at least 1 subcommand")
+	}
+
+	subcommand := args[0]
+	cmdArgs := args[1:]
+	generator, err := getCmdGenerator(subcommand)
+	if err != nil {
+		return nil, fmt.Errorf("validating command: %w", err)
+	}
+
+	for _, arg := range cmdArgs {
+		if _, ok := generator.allowedOpts[arg]; !ok {
+			return nil, fmt.Errorf("invalid argument provided for '%s' command: '%s'", subcommand, arg)
+		}
+	}
+
+	return generator.generate(ctx, cmdArgs)
+}
+
+func getCmdGenerator(cmd string) (*allowedCmdGenerator, error) {
+	allowedCmdGenerators := map[string]allowedCmdGenerator{
+		"brew": {
+			allowedOpts: map[string]struct{}{
+				"outdated": {},
+				"--json":   {},
+			},
+			generate: generateBrewCommand,
+		},
+	}
```

Comment:
Sorry if I missed this on the last pass -- might as well make this a `var allowedCmdGenerators = map...` at the top so we don't have to instantiate it on every call to `getCmdGenerator`?

Reply:
oh yep! will do